### ISSUE TITLE
Queries with limit

### ISSUE THREAD

**Issue by swistak35** [t12nslookup/activerecord-advantage-adapter#1]
Given such a query

```
Param.where(ID: "SOMEIDENTIFIER").limit(1).to_sql
```

running version 0.1.1 of the adapter and active record 3.x we got:

```
SELECT TOP 1 "PARAM".* FROM "PARAM"  WHERE "PARAM"."ID" = 'SOMEIDENTIFIER'
```

which is correct.

In 0.1.5 version of the adapter, with new Arel::Visitor which was implemented, we get:

```
SELECT  TOP  'SOMEIDENTIFIER' "PARAM".* FROM "PARAM" WHERE "PARAM"."ID" = 1
```

Basically it seems that the bindings are being attached to the wrong parts of the query (`1` and `'SOMEIDENTIFIER'` are swapped).

I'm trying to debug it, but I don't know arel internals, if you have any idea, let me know.

I'm using `activerecord` in version `5.1.7`.

**Comment by t12nslookup** [comment]
Ack! In pretty certain I tested that!

Sorry.

I will destroy 0.1.5, and add in better tests at my end.

I still need to make my changes, to fix other issues, but I don't want to
break existing code!

Jon

**Comment by t12nslookup** [comment]
I've pulled 0.1.5. you'll need to downgrade inside your Gemfile for now. I will have to push my tests to this repository, but they won't work on anyone else's, as my server is a fixed ip

**Comment by t12nslookup** [comment]
I'm not getting this issue.  can you tell me which version of rails you are using, as that changes the logic?

EDIT: Okay, sorry missed your comment, you're on rails 5.1

**Comment by t12nslookup** [comment]
okay, got it.  downgraded to ruby 2.6.7, and rails 5.1.

I will work through what differences there are and add in options to fix (hopefully trivial)

**Comment by t12nslookup** [comment]
It looks like we need some work-around for rails 5.0 and 5.1 (bind variables appear to swap places)

different rails versions:

```
Loading development environment (Rails 4.2.11.3)
2.3.8 :001 > Worksheet.unscoped.where(ID: "SOMEIDENTIFIER").limit(1).to_sql
 => "SELECT  TOP 1 \"WKSHEET\".* FROM \"WKSHEET\" WHERE \"WKSHEET\".\"ID\" = 'SOMEIDENTIFIER'"

Loading development environment (Rails 5.0.7.2)
2.3.8 :001 > Worksheet.unscoped.where(ID: "SOMEIDENTIFIER").limit(1).to_sql
 => "SELECT  TOP 'SOMEIDENTIFIER' \"WKSHEET\".* FROM \"WKSHEET\" WHERE \"WKSHEET\".\"ID\" = 1"

Loading development environment (Rails 5.1.7)
2.3.8 :001 > Worksheet.unscoped.where(ID: "SOMEIDENTIFIER").limit(1).to_sql
 => "SELECT  TOP 'SOMEIDENTIFIER' \"WKSHEET\".* FROM \"WKSHEET\" WHERE \"WKSHEET\".\"ID\" = 1"

Loading development environment (Rails 5.2.8)
2.3.8 :001 > Worksheet.unscoped.where(ID: "SOMEIDENTIFIER").limit(1).to_sql
 => "SELECT  TOP 1 \"WKSHEET\".* FROM \"WKSHEET\" WHERE \"WKSHEET\".\"ID\" = 'SOMEIDENTIFIER'"

Loading development environment (Rails 6.0.5)
2.6.7 :001 > Worksheet.unscoped.where(ID: "SOMEIDENTIFIER").limit(1).to_sql
 => "SELECT  TOP 1 \"WKSHEET\".* FROM \"WKSHEET\" WHERE \"WKSHEET\".\"ID\" = 'SOMEIDENTIFIER'"

Loading development environment (Rails 6.1.6)
2.6.7 :001 > Worksheet.unscoped.where(ID: "SOMEIDENTIFIER").limit(1).to_sql
 => "SELECT  TOP 1 \"WKSHEET\".* FROM \"WKSHEET\" WHERE \"WKSHEET\".\"ID\" = 'SOMEIDENTIFIER'"

 ```

**Comment by t12nslookup** [comment]
okay, it looks like we have a "winner". I have recreated the generic `combine_bind_parameters` that rails 5.0 and 5.1 use for their Arel, so that it puts the limit in the `binds` before any other binding, which means that they pop off in the expected order.  I'm guessing that the writers were expecting a nice database like PostgreSQL, but this must have broken many different databases.

I have committed the change to the repository, but not bumped the version to rubygems yet.  if it's possible for you to test, by putting the following in your Gemfile:
```
gem 'activerecord-advantage-adapter', :git=>'https://github.com/t12nslookup/activerecord-advantage-adapter'
```
I'd appreciate it.

different rails versions now:
```
Loading development environment (Rails 4.2.11.3)
2.3.8 :001 > Worksheet.unscoped.where.not(ROWID: 1).limit(1).to_sql
 => "SELECT  TOP 1 \"WKSHEET\".* FROM \"WKSHEET\" WHERE (\"WKSHEET\".\"ROWID\" <> '1')"

Loading development environment (Rails 5.0.7.2)
2.3.8 :001 > Worksheet.unscoped.where.not(ROWID: 1).limit(1).to_sql
 => "SELECT  TOP 1 \"WKSHEET\".* FROM \"WKSHEET\" WHERE (\"WKSHEET\".\"ROWID\" <> '1')"

Loading development environment (Rails 5.1.7)
2.3.8 :001 > Worksheet.unscoped.where.not(ROWID: 1).limit(1).to_sql
 => "SELECT  TOP 1 \"WKSHEET\".* FROM \"WKSHEET\" WHERE (\"WKSHEET\".\"ROWID\" <> '1')"

Loading development environment (Rails 5.2.8)
2.3.8 :001 > Worksheet.unscoped.where.not(ROWID: 1).limit(1).to_sql
 => "SELECT  TOP 1 \"WKSHEET\".* FROM \"WKSHEET\" WHERE \"WKSHEET\".\"ROWID\" <> '1'"

Loading development environment (Rails 6.0.5)
2.6.7 :001 > Worksheet.unscoped.where.not(ROWID: 1).limit(1).to_sql
 => "SELECT  TOP 1 \"WKSHEET\".* FROM \"WKSHEET\" WHERE \"WKSHEET\".\"ROWID\" <> '1'"

Loading development environment (Rails 6.1.6)
2.6.7 :001 > Worksheet.unscoped.where.not(ROWID: 1).limit(1).to_sql
 => "SELECT  TOP 1 \"WKSHEET\".* FROM \"WKSHEET\" WHERE \"WKSHEET\".\"ROWID\" <> '1'"

Loading development environment (Rails 7.0.3)
2.7.3 :001 > Worksheet.unscoped.where.not(ROWID: 1).limit(1).to_sql
 => "SELECT  TOP 1 \"WKSHEET\".* FROM \"WKSHEET\" WHERE \"WKSHEET\".\"ROWID\" <> '1'"
```

**Comment by swistak35** [comment]
Yes, the issue with `TOP` is fixed.

---

I have found though another one (I think, possibly I could do someting better).

Currently I had something like this:
```
class Param
  scope :with_row_id, -> { select("\"#{self.table_name}\".*, \"#{self.table_name}\".ROWID") }
end

```

To have a way to include `ROWID` together with records.

And now, `Param.count` works correctly, `Param.with_row_id.first` works correctly (`.first` as an example of some query on this scope), however `Param.with_row_id.count` does not.

Previously generated SQL was:
```
SELECT COUNT(*) FROM "AGENTVEE_ACT"
```

Now it is:
```
SELECT COUNT("AGENTVEE_ACT".*, "AGENTVEE_ACT".ROWID) FROM "AGENTVEE_ACT"
```

Same ruby, rails versions as in original issue.

For now, I've changed my code to be careful and not execute `with_row_id` when I am doing `.count` later. **With that change, and one else I will describe in another issue, all my current tests which use advantage adapter are passing :)**

---

Of course, doing `Param.with_row_id.count` on it's own doesn't make sense, but the `.with_row_id` was added at some point when building a scope in a separate method and then later used either to do the `.count` or to do `.first` or `.to_a` for example.

```
# note that this method returns a Relation, not the actual records
def some_method_building_scope
  Param
    .with_row_id
    # other scopes or filtering, wheres etc.
    .with_some_other_scopes
end 


def some_other_method
  some_method_building_scope.count
  # ...
  some_method_building_scope.to_a
end
```

**Comment by t12nslookup** [comment]
Yes, I have exactly the same issue with counts, for some reason, which is why all of my above examples work with `unscoped` in them.

Thanks for testing the updated "TOP" code.

I will investigate the `count` issue at some point, but will push these changes as they are to 0.1.6, as they are a significant improvement.

**Comment by t12nslookup** [comment]
FYI: the issue with counts is if the select statement is in quotes:
```
  default_scope { select(arel_table[:ROWID], arel_table[Arel.star]) }
```
appears to work with counts.

**Comment by t12nslookup** [comment]
ignore that comment, I think I was using an older rails for testing that didn't understand the `default_scope` method

**Comment by t12nslookup** [comment]
the only solutions for now are to use either of the following:
```
Worksheet.unscoped.where(x: 12).count
```
or:
```
Worksheet.where(x: 12).count(Arel.star)
```
(or any singular column, or a simple "*", or even just the number 1)

**Comment by t12nslookup** [comment]
I have included an "ActiveRecord::Calculations::count" in the advantage_adapter.rb, which seems to fix this issue.

in the case where there is no specified "column_name" passed to the `count` it now uses the `primary_key` from the table that the count is being called upon.

the downside of this is that if you were specifying a column with `select`, and then call a `count` (with no parameters) it would override the select parameters with the primary_key.

I chose the primary_key, rather than something like '1' for the case where:
```
Loading development environment (Rails 5.2.8)
2.3.8 :001 > Worksheet.distinct.count(1)
 => 1
2.3.8 :002 > Worksheet.count(1)
 => 6
```
